Ip k66f eth ld placement

lmajewski opened the pull request:
This patch series augments the `ip_k66f` board code to provide forced ordering of network interfaces (`struct net_if`).
As a result the output from `net_if_get_by_index()` function will be always deterministic.

Commit:
net: Append device name to linker generated symbols for struct net_if

Before this change all struct net_if objects accessed by for example
net_if_get_by_index() are placed in one linker area (i.e. net_if_area)
with the same "name" - '_net_if.static.net_if'. This may cause problems
when the order of struct net_if elements is important.

With the same names for all elements there is no guarantee of placement
order. After this change the unique device name is appended, so
SORT_BY_NAME() linker command places objects in reproductible manner.

Signed-off-by: Lukasz Majewski <lukma@denx.de>

Commit:
net: Introduce new NETWORK_RAM_SECTIONS() macro for common-ram.ld

This NETWORK_RAM_SECTIONS() macro can now be overridden (in board
specific linker.ld file) when one needs to force explicit placement
of network interface' related elements.

Signed-off-by: Lukasz Majewski <lukma@denx.de>

Commit:
arm: ip_k66f: Introduce board specific linker.ld file

This patch introduces linker.ld file specific for ip_k66f board.
It reuses the "common" linker script for other NXP SoCs with defined
NETWORK_RAM_SECTIONS() macro for explicit placement of network
interfaces on ip_k66f board.

grep -A13 net_if_area zephyr/zephyr.map
net_if_area 0x00000000200012a0 0x820 load address 0x0000000000024ac4
            0x00000000200012a0                _net_if_list_start = .
*(SORT_BY_ALIGNMENT(._net_if.static.dts*))
._net_if.static.dts_ord_51
          0x00000000200012a0 0x208 zephyr/libzephyr.a(eth_mcux.c.obj)
*(SORT_BY_NAME(SORT_BY_ALIGNMENT(._net_if.static.dsa_slave*)))
._net_if.static.dsa_slave_port_DT_N_S_soc_S_spi_4002d000_S_dsa_0_S_lan_1
          0x00000000200014a8 0x208 zephyr/libzephyr.a(dsa_ksz8794.c.obj)
._net_if.static.dsa_slave_port_DT_N_S_soc_S_spi_4002d000_S_dsa_0_S_lan_2
          0x00000000200016b0 0x208 zephyr/libzephyr.a(dsa_ksz8794.c.obj)
._net_if.static.dsa_slave_port_DT_N_S_soc_S_spi_4002d000_S_dsa_0_S_lan_3
          0x00000000200018b8 0x208 zephyr/libzephyr.a(dsa_ksz8794.c.obj)
          0x0000000020001ac0                _net_if_list_end = .

As a result the eth0 (master DSA interface) is explicitly placed as the
first one (with the 'dts_ord_51' automatically assigned name) followed
by lan{123} interfaces (with 'dsa_slave_port' name assigned in
dsa_ksz8794.c).

After this patch network interfaces are explicitly placed in correct
order, so 'net_if_get_by_index()' will work correctly.

Signed-off-by: Lukasz Majewski <lukma@denx.de>

jukkar reviewed (commented):
The changes look reasonable but there is now build failure from tests like this
```
cd /workdir/zephyr/twister-out/ip_k66f/tests/net/vlan/net.vlan/zephyr && /usr/local/bin/cmake -E echo
/opt/sdk/zephyr-sdk-0.12.3/arm-zephyr-eabi/bin/../lib/gcc/arm-zephyr-eabi/10.2.0/../../../../arm-zephyr-eabi/bin/ld: warning: orphan section `._net_if.static.eth_test' from `app/libapp.a(main.c.obj)' being placed in section `._net_if.static.eth_test'
```
also the arp and virtual tunnel tests do not pass.

lmajewski commented:
@jukkar 

The issue is a bit more complicated. To be more specific, tests rely on linker rule being `SORT_BY_NAME(._net_if.static.*)` 
When we do compile tests (which like `arp` use `net_if_get_default()` to get interface) the zephyr.map:

`net_if_area     0x0000000020000a58       0x10 load address 0x0000000000010f40`
`                0x0000000020000a58                _net_if_list_start = .`
` *(SORT_BY_NAME(SORT_BY_ALIGNMENT(._net_if.static.*)))`
` ._net_if.static.net_if`
`                0x0000000020000a58        0x8 app/libapp.a(main.c.obj)`
` ._net_if.static.net_if`
`                0x0000000020000a60        0x8 zephyr/libzephyr.a(eth_mcux.c.obj)`

And it all works as `app/libapp.a` gets linked first (as its directory is app not zephyr).

The potential solution which come to my mind would be to rename all names in `NET_DEVICE_INIT()` to for example
have `tst_` prefix (`fake_dev -> tst_fake_dev` or `eth_test_1 -> tst_eth_test1`).

Then I could add special rule for `tst_` prefixed entries for `struct net_if` objects in linker file.

jukkar commented:
> The potential solution which come to my mind would be to rename all names in `NET_DEVICE_INIT()` to for example
> have `tst_` prefix (`fake_dev -> tst_fake_dev` or `eth_test_1 -> tst_eth_test1`).

IMHO we should not depend on in which order the interfaces are in the linker list. This is just recipe for disaster later. The tests need to be tweaked a bit to not take blindly the first interface (if there are multiple) if that that 1st is not the proper one. I remember doing some changes in some tests for this.
Some (perhaps not so great) examples for this in commits like
7d90812f265c0d23bad904434cef9c1616ba08ad
865ed89b1d90ebadde4a0c1939ef45e1d6b7dff5

Commit:
test: Lookup for struct net_if interface defined in arp test

The struct net_if object placement in linker 'net_if_area' section
now depends on the order of liked libraries (like
app/libapp.a(main.c.obj). As a result interfaces defined in testing
applications are placed first in the 'net_if_area' section.

Problem arises when board has other interfaces, which are placed on the
first (i.e. default) position. In this case the test just fails as wrong
network interface is used.

Signed-off-by: Lukasz Majewski <lukma@denx.de>

Commit:
test: Rename interfaces defined in checksum offload test

The struct net_if object placement in linker 'net_if_area' section
now depends on the order of ETH_NET_DEVICE_INIT() macros in the source
code (for the same object file - like app/libapp.a(main.c.obj)).

Currently the 'eth_offloading_enabled_test' device is placed first,
followed by 'eth_offloading_disabled_test'.

This placement is important when one would like to access those
interfaces with net_if_get_default() function.

After the rename the order is explicit, when we append the device name
to the ._net_if.static.<dev name> object in linker's net_if_area.

The code now works as all struct net_if interfaces are placed as
'._net_if.static.net_if' objects.

Signed-off-by: Lukasz Majewski <lukma@denx.de>

jukkar commented:
>  one could use `net_if_lookup_by_device()` and then use the net device defined in the test.

Sure, that is one option.

Commit:
test: Rename interfaces in ptp clock test

This change is required to allow correct tests operation after changing
the definition of linker script placement rule (to be more strict and
using the device name).

The struct net_if interfaces have been renamed to preserve correct
order.
Moreover, the iface_cb() function now immediately returns for
interfaces, which are not supposed to be used in this tests - for
example ones already defined on board on which the test is run.

Signed-off-by: Lukasz Majewski <lukma@denx.de>

Commit:
test: Lookup for struct net_if interface defined in ipv6 test

The tests/net/ipv6/src/main.c uses net_if_get_default() to get the
interface for testing. This is problematic when board introduces its own
linker script, which places board's network interface on the "default"
position - i.e. the first one.

To fix this issue - the lookup for network device defined for this test
is used instead.

Moreover, interfaces rename has been performed in a manner, which would
allow working with 'eth_native_posix' interface being added as linker
object when drivers/ethernet/eth_native_posix.c file is linked for
./zephyr/tests/net/ipv6 test on 'native_posix' board.

Signed-off-by: Lukasz Majewski <lukma@denx.de>

jukkar commented on tests/net/ptp/clock/src/main.c:
```diff
@@ -285,6 +285,18 @@ static void iface_cb(struct net_if *iface, void *user_data)
+	 * The below code is to only use struct net_if devices defined in this
+	 * test as board on which it is run can have its own set of interfaces.
+	 *
+	 * As a result one will not rely on likner's specific 'net_if_area'
```
typo, s/likner/linker/

jukkar commented on drivers/ethernet/dsa_ksz8794.c:
```diff
@@ -695,17 +695,13 @@ static void dsa_delayed_work(struct k_work *item)
 
```
Please change the commit subject of "fix: dsa: ...." to "net: dsa: Initialize ...."

Commit:
test: Rename interfaces defined in virtual networks test

This work is similar to one performed for ./net/ptp/clock/ test.

However, there is one notable extension - the check in iface_cb()
callback also checks for virtual interfaces defined in
subsys/net/l2/virtual/ipip/ipip.c. Those 'virtual' interfaces are
required for this test to be run.

Signed-off-by: Lukasz Majewski <lukma@denx.de>

lmajewski commented:
@jukkar For the newest patch set (7f9c85c) I've also prepared fixes for tests, which broke down after more strict network interfaces objects linker placement.

jukkar commented on tests/net/virtual/src/main.c:
```diff
@@ -304,6 +304,19 @@ static void iface_cb(struct net_if *iface, void *user_data)
+	 * The below code is to only use struct net_if devices defined in this
+	 * test as board on which it is run can have its own set of interfaces.
+	 *
+	 * As a result one will not rely on likner's specific 'net_if_area'
```
the same typo here, s/likner/linker/

lmajewski replied:
Ok.

jukkar reviewed (commented):
The changes look good. Please fix the typos too at some point after the buildkite has finished the checks.

jukkar commented:
@all could someone elso review this?

galak commented:
@JordanYates not sure if your work has any impact here.

JordanYates commented:
@galak  I don't believe so, although a rebase would settle it. This doesn't look like it depends on `{REGION}_ADDR` or `{REGION}_SIZE`, which was the only behavioral change from my PR.
